File: Difficulty: Medium/Get Min from Stack/get-min-from-stack.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
  public:
  vector<int>s,p;
  int i=-1,j=-1;
    Solution() {
        // code here
        int MAX_Q=1e5+5;
        s.resize(MAX_Q,0);
        p.resize(MAX_Q,0);
    }

    // Add an element to the top of Stack
    void push(int x) {
        // code here
        s[++i]=x;
        if(j==-1 || s[j]>=x){
            
        
        p[i]=j;
        j=i;
        }
        
        
    }

    // Remove the top element from the Stack
    void pop() {
        // code here
        if(i==-1){
            
            return;
        }
        if(j==i){
            j=p[i];
        }
        i--;
        
        
    }

    // Returns top element of the Stack
    int peek() {
        if(i==-1){
            return i;
        }
        return s[i];
        // code here
    }

    // Finds minimum element of Stack
    int getMin() {
        // code here
        if(j==-1){
            return j;
        }
        return s[j];
    }
};


//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int q;
        cin >> q;

        Solution ob;

        while (q--) {
            int qt;
            cin >> qt;

            if (qt == 1) {
                // push
                int att;
                cin >> att;
                ob.push(att);
            } else if (qt == 2) {
                // pop
                ob.pop();
            } else if (qt == 3) {
                // peek
                cout << ob.peek() << " ";
            } else if (qt == 4) {
                // getMin
                cout << ob.getMin() << " ";
            }
        }
        cout << endl;
        cout << "~" << endl;
    }

    return 0;
}

// } Driver Code Ends

File: Difficulty: Medium/Get Min from Stack/get-min-from-stack.java
class Solution {
    int stack[],minstack[];
    int top;
    public Solution() {
        stack=new int[100001];
        minstack=new int[100001];
        top=-1;
    }

    // Add an element to the top of Stack
    public void push(int x) {
        // code here
        stack[++top]=x;
        if(top==0){
            minstack[top]=x;
        }
        else{
            minstack[top]=Math.min(minstack[top-1],x);
        }
    }

    // Remove the top element from the Stack
    public void pop() {
        if(top>=0){
            top--;
        }
        // code here
    }

    // Returns top element of the Stack
        
    public int peek() {
        // code here
        if(top==-1){
            return -1;
        }
        return stack[top];
    }

        
    // Finds minimum element of Stack
    public int getMin() {
        // code here
        
        if(top==-1){
            return -1;
        }
        return minstack[top];
    }
}